File: core/src/net/namekdev/memcop/domain/Sector.java
package net.namekdev.memcop.domain;

import java.util.Random;

/**
 * in-game memory sector
 */
public class Sector {
    public static Random random = new Random();

    /**
     * Have algorithm tried to write to it?
     */
    public boolean written = false;

    /**
     * Is this sector a bad sector?
     */
    public boolean broken = false;


    public final boolean originallyBroken;

    private int value;
    public float levelInputGradient = 0f;
    public boolean markedForGradient;


    public Sector() {
        this(false);
        reset();
    }

    public Sector(boolean originallyBroken) {
        this.originallyBroken = originallyBroken;
        reset();
    }

    public void reset() {
        written = false;
        broken = originallyBroken;
        value = broken ? random.nextInt() : 0;
        levelInputGradient = 0;
    }

    public boolean isWritable() {
        return !broken;
    }

    public void write(int value) {
        // let's ditch the value for now
        written = true;
        this.value = value;
    }

    public int value() {
        return value;
    }

    public void markForGradient(float gradient) {
        this.levelInputGradient = gradient;
        this.markedForGradient = true;
    }

    public static Sector newBroken() {
        return new Sector(true);
    }

    @Override
    public String toString() {
        if (written)
            return "_";
        if (broken)
            return "x";

        return " ";
    }
}

File: core/src/net/namekdev/memcop/domain/Sector.kt
package net.namekdev.memcop.domain

import net.namekdev.memcop.domain.pojo.SectorInfo
import java.util.Random

/**
 * in-game memory sector
 */
class Sector(val originallyBroken: Boolean) {

    /**
     * Have algorithm tried to write to it?
     */
    var written = false

    /**
     * Is this sector a bad sector?
     */
    var broken = false

    private var value: Int = 0
    var levelInputGradient = 0f
    var markedForGradient: Boolean = false

    val isWritable: Boolean
        get() = !broken

    init {
        reset()
    }

    fun reset() {
        written = false
        broken = originallyBroken
        value = if (broken) random.nextInt() else 0
        levelInputGradient = 0f
    }

    fun write(value: Int) {
        // let's ditch the value for now
        written = true
        this.value = value
    }

    fun value(): Int {
        return value
    }

    fun markForGradient(gradient: Float) {
        this.levelInputGradient = gradient
        this.markedForGradient = true
    }

    override fun toString(): String {
        if (written)
            return "_"

        return if (broken) "x" else " "
    }

    companion object {
        val random = Random()

        fun fromPojo(info: SectorInfo): Sector {
            return Sector(info.isBroken)
        }
    }
}
